README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# regcensusAPI

<!-- badges: start -->

<!-- badges: end -->

The goal of regcensusAPI is to provide an easy way for researchers to
access regulatory restrictions data from the Mercatus Center at George
Mason University, RegData. RegData quantifies the volume of regulatory
restrictions in various jurisdictions, from national to subnational. For
more information about RegData, visit
[QuantGov](https://www.quantgov.org) website.

## Installation

You can install the released version of regcensusAPI from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("regcensusAPI")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("QuantGov/regcensus-api-client-R")
```

## Usage

Using the API is easy. First, you would want to find out what topics are
covered.

``` r
library(regcensusAPI)
#> Welcome to RegCensus API - Pulling initial data
## basic example code
get_topics()
```

All data are further grouped into *series*. The function
**get\_series\_period()** returns the list of data series, the
jurisdictions, and the periods for which data are available.

``` r
get_series_period()
```

For more information on how to fully utilize regcensusAPI to obtain
regulatory restrictions data, check out the vignette accompanying the
package.


NEWS.md
# regcensusAPI 0.1.1

* Added a `NEWS.md` file to track changes to the package.
Newly released package. Please send comments and feedback to the author.


tests/testthat.R
library(testthat)
library(regcensusAPITest)

test_check("regcensusAPITest")


tests/testthat/test-data_frames.R
context("test-data_frames")

test_that("multiplication works", {
  expect_equal(2 * 2, 4)
})


R/zzz.R
.onLoad <- function(libname,pkgname){

  #packageStartupMessage("Welcome to RegCensus API - Pulling initial data")
  R.cache::saveCache(object = get_series_period(),
                     key = list(c("seresYearID","summary",
                                  "periodStartDate","periodEndDate",
                                  "periodCode","earliestDate","latestDate",
                                  "earliestPeriodCode","latestPeriodCode",
                                  "recordsAvailable","series.seriesID",
                                  "series.seriesName","series.topic.topicID",
                                  "jurisdiction.jurisdictionID",
                                  "documentSubtype.documentSubtypeID",
                                  "documentSubtype.subtypeName")),
                     compress = T)

  #To Do: write function to get all industries, regardless of jurisdiction
  R.cache::saveCache(object = get_industries(38),
                     key = list(c("industryID","jurisdiction","industryCode",
                                  "industryName","industryDescription",
                                  "industryCodeVersion","industryCodeStandard")),
                     compress = T)
}
.onAttach <- function(libname, pkgname){

}


R/values.R
# RETURN VALUES FROM USER REQUESTS

get_geoCodes <- function(jurisdictionIDs) {
    geoCodes <- c()
    i <- 1
    for (jid in jurisdictionIDs) {
        jurisdiction <- get_jurisdictions(jid)
        geoCodes[i] <- paste0(jurisdiction$geoCode)
        i <- i + 1
    }

    return(geoCodes)
}

get_seriesCodes <- function(seriesIDs) {
    seriesCodes <- c()
    i <- 1
    for (sid in seriesIDs) {
        ## Note get_series does not work correctly right now
        series <- get_series(sid)
        seriesCodes[i] <- paste0(series$x)
        i <- i + 1
    }

    return(seriesCodes)
}

get_jurisdictionID <- function(geoCodes) {
    # this function can be expanded to return vector of jids for state names, state abbr, and countries
    jurisdictions <- get_jurisdictions()
    jids <- c()
    i <- 1
    for (geoCode in geoCodes) {
        jid <- unique(jurisdictions$jurisdictionID[jurisdictions$geoCode$geoCode == geoCode])
        jids[i] <- paste(jid, collapse=",")
        i <- i + 1
    }

    return(jids)
}

get_years <- function(seriesCode) {
    # Idea is to return vector of available years for provided seriesCode and jurisdiction, if provided

}


R/helpers.R
# HELPER FUNCTIONS
## Get base url
get_baseURL <- function() {
    return("https://api.quantgov.org")
}

## Generate API call from controller and requested ID
regdata_json_request <- function(controller, id) {
    base_url <- get_baseURL()

    if (!is.na(id)) {
        apicall <- paste0(base_url, "/", controller, "/", id)
        print(apicall)
    } else {
        apicall <- paste0(base_url, "/", controller)
        print(apicall)
    }
    json <- jsonlite::fromJSON(apicall) #8/26/2019, Kofi: added full path to fromJSON
    return(json)
}

## Return URL of API call
regdata_json_requestURL <- function(controller, id) {
    base_url <- get_baseURL()

    if (!is.na(id)) {
        apicall <- paste0(base_url, "/", controller, "/", id)
    } else {
        apicall <- paste0(base_url, "/", controller)
    }

    return(apicall)
}


R/data_frames.R
# RETURN DATA FRAMES FROM USER REQUESTS



#' Return list of topics in RegCensus
#'
#' @param topic Integer - the ID of the topic of interest
#'
#' @return dataframe
#' @export
#'
#' @examples
#' get_topics(1)
#' get_topics()
get_topics <- function(topic = NA) {

  #load all topics
  if (!is.na(topic))
    url_compose <- paste0(get_baseURL(),"/topics",paste0(topic,collapse = ","))
  else
    url_compose <- paste0(get_baseURL(),"/topics")

  return(make_api_call(url_compose))
}



#' Generate a list of jurisdictions in RegData
#'
#' @param jurisdiction Integer - the ID of the jurisdiction of interest
#'
#' @return List of jurisdictions
#' @export
#' @examples
#' get_jurisdictions()
#' get_jurisdictions(38)

get_jurisdictions <- function(jurisdiction = NA) {
  if (!is.na(jurisdiction)) {
    url_compose <- paste0(get_baseURL(),"/jurisdictions", jurisdiction)
  }
    else {
      url_compose <- paste0(get_baseURL(),"/jurisdictions")

    }

    return(make_api_call(url_compose))
}

#' Return a list of industries with data in RegCensus
#'
#' @param jurisdiction An integer, the jurisdiction ID
#'
#' @return data frame with industries
#' @export
#'
#' @examples
#' get_industries(38)
get_industries <- function(jurisdiction=NA){

    if (is.na(jurisdiction)) {
        print("You must specify a jurisdiction. Select from: ")
        print(list_jurisdictions())
        stop()
    }

  url_compose <- paste0(get_baseURL(),"/industries?jurisdiction=",jurisdiction)

  return(make_api_call(url_compose))
}

#' Return a dataframe of all agencies in a jurisdiction
#'
#' @param jurisdiction Integer - the IDs of the jurisdiction of interest
#'
#' @return dataframe
#' @export
#'
#' @examples
#' get_agencies(38)
get_agencies <- function(jurisdiction = NA) {
    if (!is.na(jurisdiction)) {
        id_str <- paste(jurisdiction, collapse = ",")
        url_compose <- paste0(get_baseURL(),"/agencies/jurisdiction?jurisdictions=",id_str)
      } else {
      url_compose <- paste0(get_baseURL(),"/agencies")
    }


    return(make_api_call(url_compose))
}

#' Return a dataframe with the series for the id specified
#'
#' @param id Integer - ID for the by parameter
#'
#' @return dataframe
#' @export
#'
#' @examples
#' get_series(bidy=1)
get_series <- function(id = NA) {
  if(is.na(id)){
    url_compose <- paste0(get_baseURL(),"/series")
  } else {
    url_compose <- paste0(get_baseURL(),"/series/",id)
  }

  return(make_api_call(url_compose))
}

#' Return dataframe with all jurisdictions-series-years of data available
#'
#' @param jurisdiction Integer - the ID of the jurisdiction of interest
#'
#' @return dataframe
#' @export
#'
#' @examples
#' get_series_period()
#' get_series_period(38)
get_series_period <- function(jurisdiction=NA) {

  #if empty jurisdiction, return data for all
    if (!is.na(jurisdiction)) {
      jur_str <- paste0(find_jurisdiction(jurisdiction), collapse = ",")
      url_compose <- paste0(get_baseURL(),"/periods","?jurisdiction=",jur_str)
    } else {
      url_compose <- paste0(get_baseURL(),"/periods/available")
    }


    return(make_api_call(url_compose))
}


#' Return dataframe with values of series of interest
#'
#' @param jurisdiction Integer - the ID of the jurisdiction of interest
#' @param series Series ID
#' @param time The time
#' @param agency The agency ID
#' @param industry The industry code using the jurisdiction-specific coding system
#' @param date_is_range Boolean indicating whether the time parameter is range or should be treated as single data points
#' @param summary Boolean - Return summary instead of document level data
#' @param filtered Boolean - Exclude poorly-performing industry classification results
#' @param documentType Integer - ID of document
#' @param document Integer - List of document IDs to return data for. Use when requesting document-level data instead of
#' summary data.
#' @param include_metadata Boolean - Should dataframe include the metadata to make it more readable
#'
#' @return dataframe
#' @export
#'
#' @examples
#' get_values(jurisdiction = 38, series = c(92), time = c('1990','2000'),
#' industry = c('111','33'), agency = c(66,111))

get_values <- function(jurisdiction, series, time=c(2015,2019), summary=TRUE,
                       filtered=TRUE, documentType=3, agency=0, industry='0',
                       date_is_range = TRUE, document=NA, include_metadata=FALSE) {

    if (!summary) {
      print("Request for document-level data could take a while to download. Please be patient.")
    }


    series_str <- paste(series, collapse = ",")

    time_str <- paste(time, collapse = ",")

    industry_str <- paste(industry,collapse = ",")

    agency_str <- paste(agency,collapse = ",")

    if (length(series) > 0) {
        url_compose <- paste0(get_baseURL(),"/values?series=",series_str)
    }else
    {
        print("Valid series ID required. Select from the following list:")
        print(list_series())
        stop("Processing terminated.")

    }
    if (length(jurisdiction) > 0) {
      if (!is.numeric(jurisdiction[1])) {
        jur_str <- paste(find_jurisdiction(jurisdiction),collapse = ",")
      }else{
        jur_str <- paste(jurisdiction, collapse = ",")
      }

      url_compose <- paste0(url_compose,"&jurisdiction=",jur_str)
    }else
        stop("Jurisdiction is required.")

    if (length(agency) > 0) {
        url_compose <- paste0(url_compose,"&agency=",agency_str)
    }

    if (length(industry) > 0) {
        url_compose <- paste0(url_compose,"&industry=",industry_str)
    }

    if (length(time) > 0) {
        url_compose <- paste0(url_compose,"&time=",time_str,"&date=",time_str)
    }else{
        stop("Time is required.")
    }

    if (date_is_range) {
        url_compose <- paste0(url_compose,"&dateIsRange=", date_is_range)
    }

    if (filtered) {
      url_compose <- paste0(url_compose,"&filteredOnly=","true")
    }

    if(summary){
      url_compose <- paste0(url_compose,"&summary=","true")
    } else {
      url_compose <- paste0(url_compose,"&summary=","false")
    }
    url_compose <- paste0(url_compose,"&documentType=",documentType)

    return(make_api_call(url_compose, TRUE))
}

#' Return values for a list of jurisdictions-series-year combination. It will return data for both
#' national and sub-national (for example, state/province) within the jurisdiction.
#'
#' @param jurisdiction Integer - Jurisdiction IDs for the data
#' @param series Integer - Series IDs
#' @param time Date string - For period of interest
#' @param industries Text - List of industries to pull data for using the jurisdiction specific coding system
#' @param date_is_range Boolean - Indicate if the years values is range instead of just single values
#' @param filtered Boolean - Indicate whether to return filtered data, i.e., the industry-relevant data with poorly classified industries excluded
#'
#' @return dataframe
#' @export
#'
get_country_values <- function(jurisdiction = c(38,75), series = c(1,2),
                             time = c(2010,2011), industries = NA,
                             date_is_range=TRUE, filtered=TRUE){
    if (length(time) == 0) {
      stop("You need to include valid time period")
    }else
        time_str <- paste0(time, collapse = ",")

    if (length(jurisdiction) == 0) {
        print("You need to select at least one of the following jurisdiction IDs")
        print(list_jurisdictions())
        stop()
    }else{
        jurisdiction_str <- paste0(jurisdiction, collapse = ",")
        url_compose <- paste0(get_baseURL(),"/values/country?countries=",jurisdiction_str,"&years=",time_str)
    }

    if (length(series) == 0) {
        print("You need to select at least one of the following series IDs")
        print(list_series())
        stop()

    }else{
        series_str <- paste0(series, collapse = ",")
        url_compose <- paste0(url_compose, "&series=",series_str)
    }

    if (length(industries) == 0) {
        industries_str <- '0'
    }else{
      industries_str <- paste0(industries, collapse = ",")
        url_compose <- paste0(url_compose, "&industries=",industries_str)
    }

    if (date_is_range == FALSE)
        url_compose <- paste0(url_compose,"&dateIsRange=", "false")
    else{
      url_compose <- paste0(url_compose,"&dateIsRange=", "true")
    }

    if (filtered == FALSE)
      url_compose <- paste0(url_compose,"&filteredOnly=",filtered)


    return(make_api_call(url_compose))

}


#' Title Return the list of documents for a series-jurisdiction combination
#'
#' @param document_type Integer - ID of document type of interest. Obtain from get_document_types().
#' @param jurisdiction Integer - ID of jurisdiction of interest
#'
#' @return data frame
#' @export
#'
#' @examples
#' get_documents(55,3)
get_documents <- function(jurisdiction, document_type=3){
  print("Document-level data could take a while to download. Please be patient.")
  if (!is.null(jurisdiction)) {
    jur_str <- paste(jurisdiction, collapse = ",")
    url_compose <- paste0(get_baseURL(),"/documents?jurisdiction=", jur_str)
  }else{

    print("Please select jurisdictions from the list below:")
    list_jurisdictions()
    stop("Jurisdiction is required.")

  }

  if (length(document_type) > 0) {
    url_compose <- paste0(url_compose, "&documentType=",document_type)
  }
  else {
    print("Document type (documentType) is required. Select a document type from the list below:")
    print(get_document_types(NULL))
    stop("Invalid document type specified.")
  }


   documents <- make_api_call(url_compose)
   if (length(documents) > 0) {
     return (documents)
  }else{
    stop(paste0("No data found.",url_compose))
  }

}

#' Title Return the list of document types for a jurisdiction/or all jurisdictions.
#'
#' @param jurisdiction Integer - ID for the jurisdiction of interest. If NA, return all jurisdictions
#'
#' @return data frame of document types
#' @export
#'
#' @examples
#' get_document_types(38)
#' get_document_types()
get_document_types <- function(jurisdiction=NA){

  if (!is.na(jurisdiction)) {
    id_str <- paste(jurisdiction, collapse = ",")
    url_compose <- paste0(get_baseURL(),"/documenttypes?jurisdictions=",id_str)
  } else {
    url_compose <- paste0(get_baseURL(),"/documenttypes")
  }


  json <- make_api_call(url_compose)
  return(json)
}

#' Return series values for a set of industry codes (using NAICS)
#'
#' @param jurisdiction An Integer - jurisdiction(s) of interest. Obtain list of jurisdictions from list_jurisdictions()
#' @param series Integer -  (List of ) Series of interest. Obtain valid list from list_series(id, by)
#' @param time Date (string) - String format of dates. For summary data, just the year is enough. For daily data, use full date format such as
#' '2018-10-12'
#' @param agency Integer - List of agencies. Obtain from get_agencies()
#' @param industry_type String - List of industry code types desired. Valid values are "all", "2-Digit","3-Digit","4-Digit","5-Digit","6-Digit"
#' @param date_is_range Boolean - Date parameter is range
#' @param filtered_only Boolean - For industry values, include only good-performing industry classifications.
#' @param document_type Integer - The type of document. Obtain from list_document_types()
#' @param summary Boolean - Summary data
#'
#' @return Data frame
#' @export
#'

get_industry_values <- function(jurisdiction=c(38),
                                industry_type=NA,
                                industry=NA,
                                series=1,
                                time=c('2015','2019'),
                                agency=c(0),
                                date_is_range=TRUE,
                                filtered_only=TRUE,
                                summary=TRUE,
                                document_type=1) {
  #list of industry options
  industry_types = c("all","2-Digit","3-Digit","4-Digit","5-Digit","6-Digit")

  #parse parameters
  series_str <- paste(series, collapse = ",")
  time_str <- paste(time, collapse = ",")
  industry_str <- paste(industry_type,collapse = ",")
  agency_str <- paste(agency,collapse = ",")
  document_type_str = paste(document_type,collapse = ",")

  #if names are used, find the corresponding jurisdiction IDs
  if (length(jurisdiction) > 0) {
    if (!is.numeric(jurisdiction[1])) {
      jur_str <- paste(find_jurisdiction(jurisdiction),collapse = ",")
    }else {
      jur_str <- paste(jurisdiction, collapse = ",")
    }
    url_compose <- paste0(get_baseURL(),"/values?jurisdictions=",jur_str)
  }else{
    print("Jurisdiction is required. Select valid jurisdiction IDs from the following:")
    print(list_jurisdictions())
    stop("Invalid jurisdiction specified.")
  }

  #ensure correct industry type is selected
  if (length(industry_type) > 0) {
    # get list of industry types
    industry_types <- get_industries(jurisdiction=jurisdiction)%>%
      dplyr::mutate(ndigits = stringi::stri_length(industryCode))
    ## 2-digit
    naics2 <- industry_types %>%filter(ndigits==2)
    naics3 <- industry_types %>%filter(ndigits==2)
    naics4 <- industry_types %>%filter(ndigits==2)
    naics5 <- industry_types %>%filter(ndigits==2)
    naics6 <- industry_types %>%filter(ndigits==2)


    ## 3-digit
    url_compose <- paste0(url_compose,"&industryType=",industry_type)
  }else{
    print("Provide valid industry code type.")
    print(paste("Select valid industryType from :", industry_types))
    stop("Invalid industry type specified.")
  }


  #ensure correct series IDs are selected
  if (length(series) > 0) {
    url_compose <- paste0(url_compose,"&series=",series_str)
  }else{
    print("Provide valid series id. Select valid series IDs from the following:")
    print(get_series(by = "all"))
    stop("Invalid series ID specified.")
  }

  #ensure the correct dates are selected
  if (length(date) > 0) {
    url_compose <- paste0(url_compose,"&date=",time_str)
  } else {
    stop("Invalid dates specified")
  }

  if (date_is_range) {
      url_compose <- paste0(url_compose,"&dateIsRange=true")
  }
  else {
      url_compose <- paste0(url_compose,"&dateIsRange=false")
    }

  if (summary) {
    url_compose <- paste0(url_compose,"&summary=true")
  } else {
    url_compose <- paste0(url_compose,"&summary=false")
  }

  if (filtered_only) {
    url_compose <- paste0(url_compose,"&filteredOnly=true")
  } else {
    url_compose <- paste0(url_compose,"&filteredOnly=false")
    print("Unfiltered results are not reliable. Use at your own discretion.")
  }

  if (length(document_type) > 0) {
     #dt <- get_document_types()
       url_compose <- paste0(url_compose,"&documentType=",document_type_str)


  } else {
    print("Select valid Document Type (documentType) from the following list:")
    list_document_types()
  }

  if (length(agency) > 0) {
     url_compose <- paste0(url_compose,"&agencies=",agency_str)
  }

  return(make_api_call(url_compose))

}


R/list.R
# LIST TOPICS, SERIES, AND JURISDICTIONS

#' Generate a list of Topics in RegData
#'
#' @param topicID An integer.
#' @return List of topics and their corresponding IDs in the console.
#' @export
#' @examples
#' list_topics()
#' list_topics(1)
list_topics <- function(topicID = NA) {

    print(R.cache::loadCache(key = list(c("industryID","jurisdiction","industryCode",
                                "industryName","industryDescription",
                                "industryCodeVersion","industryCodeStandard"))))
    json <- regdata_json_request("topics", topicID)

    if (length(json) > 0) {
        topic_text <- paste0(json$topicName, " (ID: ", json$topicID, ")")
    } else {
        topic_text <- "No topic found"
    }
    return(topic_text)
}

#' Generate a list of Jurisdictions in RegData. This function is helpful for identifying the jurisdictions covered and their identifiers used in the RegData system. Note that the function accepts one jurisdictionID, distinct from FIPs or geoCodes.
#'
#' @param jurisdictionID An integer.
#' @return List of jurisdictions and their corresponding IDs and geoCodes in the console.
#' @export
#' @examples
#' list_jurisdictions()
#' list_jurisdictions(38)
list_jurisdictions <- function(jurisdictionID = NA) {
    json <- regdata_json_request("jurisdictions", jurisdictionID)
    if (length(json) > 0) {
        jurisdiction_text <- paste0(json$jurisdictionName, ", ", json$country,
                                    " (jurisdictionID: ", json$jurisdictionID,
                                    "; geoCode: ", json$geoCode$geoCode,
                                    ")")
    } else {
        jurisdiction_text <- "No jurisdiction"
    }
    return(jurisdiction_text)
}

#' Generate a list of Agencies in RegData. Note that the function accepts a vector of multiple jurisdictionIDs, distinct from FIPs or geoCodes.
#'
#' @param jurisdictionIDs An integer or vector of integers.
#' @return List of agencies and their corresponding IDs and geoCodes in the console.
#' @export
#' @examples
#' list_agencies()
#' list_agencies(38)
#' list_agencies(c(33, 38))
list_agencies <- function(jurisdictionIDs = NULL) {
    ## Return agencies available in RegData (allowed to pass multiple jurisdiction IDs to get agencies by jurisdiction)
    ## I use NULL rather than NA for this function to deal with case of multiple ids being passed.
    ## if (is.na(c(10,NA))) returns FALSE because it only checks first value.
    if (!is.null(jurisdictionIDs)) {
        id_str <- paste(jurisdictionIDs, collapse = ",")
        json <- regdata_json_request(paste0("agencies/jurisdiction?jurisdictions=", id_str), NA)
    } else {
        json <- regdata_json_request("agencies", NA)
    }

    if (length(json) > 0) {
        agency_text <- paste0(json$agencyName, ", "," (agencyID: ", json$agencyID, ")")
    } else {
        agency_text <- "No jurisdiction"
    }
    return(agency_text)

}

#' Generate a list of Series in RegData and their corresponding series IDs and series codes.
#'
#' You can obtain information on series by:
#' \itemize{
#' \item series ID (default). Only one series ID is accepted. Lists by series do not require an id to be defined.
#' \item series code(s). Multiple series codes can be selected by passing a vector of strings as the "id" parameter.
#' \item agency or agencies. The function accepts multiple agency IDs passed as a vector.
#' \item jurisdictions. The function accepts jurisdiction IDs (distinct from geoCodes) as a vector.
#' \item topic. Only one topic ID is allowed.
#' }
#' No "id" is required when selecting series by series. See first example. When selecting series by series codes, agencies, jurisdictions, or topic, the "id" must be specified.
#'
#' @param id An integer or vector of integers. See description above and examples below for when vectors are allowed and when single values are required.
#' @param by A string specifying what type of identifier you are trying to pass in the "id" parameter. Default is "series," and accepted values are "seriesCodes", "agencies", "jurisdictions", and "topic".
#' @return List of series and their corresponding IDs and series codes in the console.
#' @export
#' @examples
#' list_series()

list_series <- function(id = NA, by = c("all", "series", "agencies", "industries", "jurisdictions", "topics")) {

    by <- match.arg(by)

    ## check length of id argument to make it sure it is compatible with selected series filter (write error messages later)
    id_len <- length(id)

    if (by == "all") {

        url <- regdata_json_requestURL("series", id)
        json <- regdata_json_request("series", id)

    } else if (by == "series") {

        if (is.na(id))
            stop("You must provide the series in the 'id' parameter.")

        # more than one seriesCode can be passed, so collapse into comma-separated list
        idstr <- paste(id, collapse = ",")

        url <- regdata_json_requestURL(paste0("series/seriesCode?series=", idstr), id = NA)

        json <- regdata_json_request(paste0("series/seriesCode?series=", idstr), id = NA)

    } else if (by == "agencies") {

        if (is.na(id))
            stop("You must provide the agency in the 'id' parameter.")
        # more than one agency can be passed, so collapse into comma-separated list
        idstr <- paste(id, collapse = ",")
        url <- regdata_json_requestURL(paste0("series/agencies?agencies=", idstr), id = NA)
        json <- regdata_json_request(paste0("series/agencies?agencies=", idstr), id = NA)

    } else if (by == "jurisdictions") {
        # more than one geoCode can be passed, so collapse into comma-separated list
        if (is.na(id))
            stop("You must provide jurisdiction in the 'id' parameter.")
        idstr <- paste(id, collapse = ",")
        url <- regdata_json_requestURL(paste0("series/jurisdiction?jurisdictions=", idstr), id = NA)
        json <- regdata_json_request(paste0("series/jurisdiction?jurisdictions=", idstr), id = NA)
    } else {
        if (is.na(id))
            stop("You must provide the topic in the 'id' parameter.")
        url <- regdata_json_requestURL("series/topics", id)
        json <- regdata_json_request("series/topics", id)
    }

    if (length(json) > 0) {
        series_text <- paste0(json$seriesName, " (Series ID: ", json$seriesID, "; Series Code: ", json$seriesCode, ")")
    } else {
        series_text <- "No series found"
    }

    return(series_text)
}

#' Generate a list of Start Years, End Years, and Total Years Available for Series by Jurisdiction in RegData.
#'
#' Each item in the list returns:
#' \itemize{
#' \item Start Year
#' \item End Year
#' \item Number of Years Available
#' \item Series ID
#' \item Series Code
#' \item Jurisdiction ID
#' \item Geo Code
#' }
#'
#' @param jurisdiction An integer. Only one jurisdiction value may be passed as an argument.
#' @return List of series with information for each series regarding start year, end years, and total years available, by jurisdiction
#' @export
#' @examples
#' list_seriesyear(10)
list_seriesyear <- function(jurisdiction = NA) {

    if (!is.na(jurisdiction)) {
        jur_str <- paste0(jurisdiction, collapse = ",")
        json <- regdata_json_request(paste0("periods/?jurisdictionID=", jur_str), NA)

    } else {
        stop("Jurisdiction required.")
    }

    if (length(json) > 0) {
        seriesyear_text <- paste0(json$series$seriesName, " in ", json$jurisdiction$jurisdictionName,
                                  " (startYear: ", json$startYear,
                                  "; endYear: ", json$endYear,
                                  "; yearsAvailable: ", json$yearsAvailable,
                                  "; seriesID: ", json$series$seriesID,
                                  "; seriesCode: ", json$series$seriesCode,
                                  "; jurisdictionID: ", json$jurisdiction$jurisdictionID,
                                  "; geoCode: ", json$jurisdiction$geoCode$geoCode,
                                  ")")
    } else {
        seriesyear_text <- "No series years found"
    }

    return(seriesyear_text)
}


#' List the document types available for a jurisdiction or all jurisdiction
#'
#'
#' @return Data frame with the list of document types available
#' @export
#'
#' @examples
#' list_document_types()
list_document_types <- function(){
        json <- regdata_json_request("documenttypes", NA)

        if (length(json) > 0) {
            document_text <- paste0(json$subtypeName, " (ID: ", json$documentSubtypeID, ")")
            json <- regdata_json_request("documenttypes", NA)
        } else {
            document_text <- "No document found"
        }
        return(document_text)

}
